Update update strategy docs (#1883)
elastic/cloud-on-k8s: docs/elasticsearch-spec.asciidoc

[id="{p}-update-strategy"]
=== Update strategy

> Users can choose to constrain the operator on the resources it can use when updating using `updateStrategy' like below.

Reviewer: How about:

> You can use `updateStrategy' to constrain the operator on the resources it can use.

Author: Ah, sorry, I forgot to push one change from https://github.com/elastic/cloud-on-k8s/pull/1883#discussion_r332007714, it was changing the first paragraph to the following:

```
Sometimes the operator must take a Pod down, for example, to restart a node and apply a new configuration value. Other times the operator must spin up pods above what is currently in the spec, say to migrate to a new node set. You can limit the number of these simultaneous changes  using an `updateStrategy` as follows:
```

Should we go with the below anyway?
```
You can use `updateStrategy' to constrain the operator on the resources it can use.
```

Reviewer: If it doesn't change the meaning of the overall sentence, then yes.

[source,yaml]
----
spec:
  updateStrategy:
    changeBudget:
      maxSurge: 3
      maxUnavailable: 1
----

> `maxSurge` refers to the number of Pods that can be scheduled above the total number of Pods in currently applied spec.

Reviewer 2: ```suggestion
`maxSurge` refers to the number of Pods that can be scheduled above the total number of Pods in the currently applied spec.
```

> `maxUnavailable` refers to the number of Pods that can be unavailable out of the total number of Pods in currently applied spec. Unavailable is defined as Pod being not ready from Kubernetes perspective.

Reviewer 2: ```suggestion
`maxUnavailable` refers to the number of Pods that can be unavailable out of the total number of Pods in the currently applied spec. Unavailable is defined as Pod being not ready from Kubernetes perspective.
```

> `maxUnavailable` refers to the number of Pods that can be unavailable out of the total number of Pods in the currently applied spec. Unavailable is defined as Pod being not ready from Kubernetes perspective.

Reviewer: Possible rewording:

> `maxSurge` refers to the number of Pods that you can schedule in addition to the total number of Pods.
>  `maxUnavailable` refers to the number of Pods that can be unavailable out of the total number of Pods. Pods are defined as unavailable when they are not ready from a Kubernetes perspective.

Not sure if it's really needed to repeat "in the currently applied spec", it's probably intuitive enough as the text refers to the code sample above.

Author: I've added it to distinguish between total number of actual (currently running) pods and total number of pods in the spec (not necessarily created yet). I think that without being specific, it's not immediately clear which one is it. WDYT?

Reviewer: How about this:

> `maxSurge` refers to the number of Pods that you can schedule in addition to the total number of Pods that are currently running.
> `maxUnavailable` refers to the number of Pods that can be unavailable out of the total number of Pods that are not created yet. Pods are defined as unavailable when they are not ready from a Kubernetes perspective.

> Note that these constraints are enforced immediately after spec is applied, which often means that the actual state of the cluster is in violation of these constraints. This is expected and the operator won't stop from moving the cluster towards respecting them.

Author: By using 'often' I tried to convey that it is a possibility and can (but doesn't need to) happen. Maybe going with 'might' would be better. WDYT?

> Note that these constraints are enforced immediately after spec is applied, which often means that the actual state of the cluster is in violation of these constraints. This is expected and the operator won't stop from moving the cluster towards respecting them.

Reviewer 3: Yes, 👍 with "might"

> Note that these constraints are enforced immediately after spec is applied, which might mean that the actual state of the cluster is in violation of these constraints. This is expected and the operator won't stop from moving the cluster towards respecting them.

Reviewer 2: ```suggestion
Note that these constraints are enforced immediately after the spec is applied, which might mean that the actual state of the cluster is in violation of these constraints. This is expected and the operator won't stop from moving the cluster towards respecting them.
```

> For example, if a new spec defines a larger cluster with no unavailable Pods permitted (`maxUnavailable: 0`), the operator will continue creating missing Pods according to the best practices. Similarly, if a new spec defines a smaller cluster with no Pods above the spec permitted (`maxSurge: 0`), the operator will continue safely removing unnecessary Pods.

Reviewer: Reworded as discussed:

> As these constraints are enforced immediately after the specification is applied, it is possible the the actual state of the cluster might violate these constraints. As long as it remains within the predefined constraints defined by `maxSurge` and `maxUnavailable`, the operator allows the cluster to scale up or down.
> For example, if a new specification defines a larger cluster with `maxUnavailable: 0`, the operator creates the missing Pods according to the best practices. Similarly, if a new specification defines a smaller cluster with `maxSurge: 0`, the operator safely removes the unnecessary Pods.

Reviewer: After discussing this offline, here is the proposed rewording:

> As these constraints are enforced immediately after the specification is applied, it is possible the the actual state of the cluster might violate these constraints.
> 
> When the cluster is outside these constraints, the operator can add or remove nodes only with the purpose of getting closer to these constraints.
> 
> When the cluster is within these constraints, the operator can either add or remove nodes, as long as the cluster remains within the predefined constraints.
> 
> For example, if a new specification defines a larger cluster with maxUnavailable: 0, the operator creates the missing Pods according to the best practices. Similarly, if a new specification defines a smaller cluster with maxSurge: 0, the operator safely removes the unnecessary Pods.

==== Specifying changeBudget
For both `maxSurge` and `maxUnavailable` you can specify the following values:

* `null` - the default value will be used,
* non-negative - the value will be used as is,

> * negative - the value will be unbounded.

Reviewer: * `null` - the default value is used
* non-negative - the value is used as is
* negative - the value is unbounded

==== Default behavior

> When `updateStrategy` is not present in the spec, it defaults to the following:

Reviewer: spec => specification

Author: Should I rename all `spec`/`specs` to `specification`/`specifications`?

Reviewer: It would be good

[source,yaml]
----
spec:
  updateStrategy:
    changeBudget:
      maxSurge: -1
      maxUnavailable: 1
----
`maxSurge` is unbounded. This results in immediate creation of all missing Pods that best practices allow for at the given time.

> `maxUnavailable` defaults to `1`. This assures that cluster will have at most one unavailable Pod at any given time.

Reviewer: `maxSurge` is unbounded: This means that all the missing Pods are immediately created according to the best practices at any given time.
  `maxUnavailable` defaults to `1`: This assures that clusters have at most one unavailable Pod at any given time.

==== Caveats

> * Setting both `maxSurge` and `maxUnavailable` to `0` will prevent the operator from performing any update that needs to bring down or create a new Pod.

Reviewer: With both `maxSurge` and `maxUnavailable` set to `0`, the operator cannot bring down an existing Pod nor create a new Pod.

> * Due to the safety measures employed by the operator, a particular changeBudget might be insufficient to perform the requested update. For example, removing last data node from one nodeSet and adding data node to a different nodeSet is not possible with `maxSurge` set to 0. In this case operator can't remove the old node (because it can't be migrated since no other data nodes are available) and can't add the new node (because `maxSurge` is set to 0).

Reviewer: Due to the safety measures employed by the operator, a particular changeBudget might be insufficient to perform the requested update. For example, with maxSurge set to 0 you cannot remove the last data node from one nodeSet and add a data node to a different nodeSet. In this case the operator cannot remove the old node and migrate data as other nodes are not available, and can't add the new node because maxSurge is set to 0.

> * Currently, the operator has limited understanding of the cluster topology and, during update, might not pick the optimal order of Pod additions and removals. This may lead to update being stuck and require specifying higher `maxSurge` than theoretically needed.

Reviewer: Currently, the operator has a limited understanding of the cluster topology and during the update might not pick the optimal order of Pods being added or removed. This can block the update process and you might need to specify a higher `maxSurge` than theoretically needed.

> In presence of any of the above, operator will emit logs indicating that upscaling or downscaling is limited to respect `maxSurge` or `maxUnavailable` setting.

Reviewer: If any of the above occurs, the operator generates logs to indicate that upscaling or downscaling are limited by `maxSurge` or `maxUnavailable` settings.

[id="{p}-update-strategy"]
=== Update strategy

> Users can choose to constraint the operator on the resources it can use when updating using `updateStrategy' like below.

Reviewer 2: ```suggestion
You can limit the number of simultaneous changes the operator will perform using an `updateStrategy` as follows:
```

Author: Would that be specific enough? I'm worried that "changes" is too broad. WDYT?

Reviewer 2: I see what youre saying, but we do clarify what changes we mean shortly there after which might be enough?

Maybe it would make sense to add a conceptual overview before we get into specific examples? Something along the lines of: "sometimes the operator must take a pod down, say to restart a node and apply a new configuration value. Other times an operator must spin up pods above what is currently in the spec, say to migrate to a new nodeSet"

[source,yaml]
----
spec:
  updateStrategy:
    changeBudget:
      maxSurge: 3
      maxUnavailable: 1
----

> `maxSurge` refers to the number of pods that can be scheduled above the total number of pods in current spec.

Reviewer 2: That is definitely something we are not consistent about, but probably should be consistent for all native types. Not sure it needs to block this PR, we talked about it last release too :)

`maxUnavailable` refers to the number of pods that can be unavailable out of the total number of pods in current spec. Unavailable is defined as pod being not ready from Kubernetes perspective.

Note that these constraints are enforced immediately after spec is applied, which often means that the actual state of the cluster is in violation of these constraints. This is expected and operator won't stop from moving the cluster towards respecting them.

> For example, if a new spec defines a larger cluster with no unavailable pods permitted, operator will keep on creating missing pods according to the best practices. Similarly, if a new spec defines smaller cluster with no pods above the spec permitted, the operator will keep on safely removing unneeded pods.

Reviewer 2: ```suggestion
For example, if a new spec defines a larger cluster with no unavailable pods permitted (`maxUnavailable: 0`), the operator will continue creating missing pods according to the best practices. Similarly, if a new spec defines a smaller cluster with no pods above the spec permitted (`maxSurge: 0`), the operator will continue safely removing unnecessary pods.
```

> For example, if a new spec defines a larger cluster with no unavailable pods permitted, operator will keep on creating missing pods according to the best practices (link). Similarly, if a new spec defines smaller cluster with no pods above the spec permitted, the operator will keep on safely removing unneeded pods.

Reviewer 4: I would say either "the operator" or "ECK". Just "operator" sounds weird?

> For example, if a new spec defines a larger cluster with no unavailable pods permitted, operator will keep on creating missing pods according to the best practices (link). Similarly, if a new spec defines smaller cluster with no pods above the spec permitted, the operator will keep on safely removing unneeded pods.

Reviewer 4: `a smaller cluster`

==== Specifying changeBudget
For both `maxSurge` and `maxUnavailable` you can specify the following values:

> * `null` - default value will be used,

Reviewer 4: `the default value`

`maxUnavailable` defaults to `1`. This assures that cluster will have at most one unavailable pod at any given time.

==== Caveats
* Setting both `maxSurge` and `maxUnavailable` to `0` will prevent the operator from performing any update that needs to bring down or create a new pod.
* Due to the safety measures employed by the operator, a particular changeBudget might be insufficient to perform the requested update. For example, removing last data node from one nodeSet and adding data node to a different nodeSet is not possible with `maxSurge` set to 0. In this case operator can't remove the node (because it can't be migrated since no other data nodes are available) and can't add the node (because `maxSurge` is set to 0).
* Currently, the operator has limited understanding of the cluster topology and, during update, might not pick the optimal order of pod additions and removals. This may lead to update being stuck and require specifying higher `maxSurge` than theoretically needed.

In presence of any of the above, operator will emit logs similar to the following:

```
Limiting nodes creation to respect maxSurge setting   {(...) "target": 2, "actual": 0}
```

```

> Cannot downscale StatefulSet  {(...) "reason": "Not removing node to respect maxUnavailable setting"}

Reviewer 4: Not sure we should copy the logs content in the documentation, pretty sure they will change at some point and we won't update the docs accordingly.

Author: Wanted to give users something to look for in the logs, but I guess I can be less explicit and convey the same thing. Fixed.
